radix-engine/src/engine/kernel.rs

```diff
@@ -469,139 +446,346 @@ where
 
-                let node_pointer = Self::current_frame(&self.call_frames)
-                    .get_node_pointer(node_id)
-                    .unwrap();
-                references_to_add.insert(node_id, node_pointer);
-                actor
+                REActor::Function(ResolvedFunction::Scrypto {
+                    package_address,
+                    blueprint_name,
+                    ident: function_name,
+                    export_name: fn_abi.export_name.clone(),
+                    return_type: fn_abi.output.clone(),
+                    code: package.code,
+                })
             }
-            MethodIdent::Native(native_fn) => REActor::Method(ResolvedReceiverMethod {
-                receiver: receiver.clone(),
-                method: ResolvedMethod::Native(native_fn.clone()),
-            }),
-        };
-
-        Ok((actor, references_to_add))
-    }
-
-    fn resolve_function_actor(
-        &mut self,
-        function_ident: FunctionIdent,
-        input: &ScryptoValue,
-    ) -> Result<(REActor, HashMap<RENodeId, RENodePointer>), RuntimeError> {
-        let mut references_to_add = HashMap::new();
+            ScryptoFnIdent::Method(ScryptoMethodIdent {
+                receiver,
+                method_name,
+            }) => {
+                let original_node_id = match receiver {
+                    ScryptoReceiver::Global(address) => {
+                        RENodeId::Global(GlobalAddress::Component(address))
+                    }
+                    ScryptoReceiver::Component(component_id) => RENodeId::Component(component_id),
+                };
 
-        let actor = match &function_ident {
-            FunctionIdent::Scrypto {
-                package_address,
-                blueprint_name,
-                ident,
-            } => {
-                let (actor, node_id) =
-                    self.execute_in_mode(ExecutionMode::ScryptoInterpreter, |system_api| {
-                        ScryptoInterpreter::<I, W>::load_scrypto_actor(
-                            ScryptoFnIdent::Function(
-                                *package_address,
-                                blueprint_name.clone(),
-                                ident.to_string(),
-                            ),
-                            input,
-                            system_api,
-                        )
-                        .map_err(|e| match e {
-                            InvokeError::Downstream(runtime_error) => runtime_error,
-                            InvokeError::Error(error) => RuntimeError::InterpreterError(
-                                InterpreterError::InvalidScryptoFunction(
-                                    function_ident.clone(),
-                                    error,
-                                ),
-                            ),
-                        })
-                    })?;
+                // Deref if global
+                let resolved_receiver =
```

Comment:
Resolving deref shouldn't be a scrypto interpreter decision but decision made in the kernel side. Can we move this back?

Reply:
As discussed, main reason for moving deref into fn resolver is because there are already derefing happening there, e.g. package address deref. 

Easier to have all fn resoliving logic in one place, so can be moved to the interepreter as a whole later on.